jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp

```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
   return integer2string(lbl_id);
 }
 
+/// Create a temporary variable
+/// /param symbol_table: a symbol table
+/// /param prefix: prefix of the name of the temporary variable
+/// /param name: base name of the temporary variable, to be append with "_tmp"
+/// /param type: type of the temporary variable
+/// /return returns new variable
+static symbol_exprt tmp_variable(
+  symbol_tablet &symbol_table,
+  const std::string &prefix,
+  const std::string &name,
+  const typet &type)
+{
+  irep_idt base_name = name + "_tmp";
+  irep_idt identifier = prefix + "::" + id2string(base_name);
+
+  auxiliary_symbolt tmp_symbol;
+  tmp_symbol.base_name = base_name;
+  tmp_symbol.is_static_lifetime = false;
+  tmp_symbol.mode = ID_java;
+  tmp_symbol.name = identifier;
+  tmp_symbol.type = type;
+  symbol_table.add(tmp_symbol);
+
+  symbol_exprt result(identifier, type);
+  result.set(ID_C_base_name, base_name);
+  return result;
+}
+
+/// Creates a monitorenter/monitorexit code_function_callt for
+/// the given synchronization object.
+/// \param symbol_table: a symbol table
+/// \param is_enter: Indicates whether we are creating a monitorenter or exit.
+/// \param object: An expression representing object  the object that need to
+/// perform a monitorenter/monitorexit operation
+static codet get_monitor_call(
+  symbol_tablet &symbol_table,
+  bool is_enter,
+  const exprt &object)
+{
+  code_typet type;
+  type.return_type() = void_typet();
+  type.parameters().resize(1);
+  type.parameters()[0].type() = java_reference_type(void_typet());
+  std::string symbol;
+
+  if(is_enter)
+    symbol="java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V";
+  else
+    symbol="java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V";
+
+  symbol_tablet::symbolst::const_iterator it
+    = symbol_table.symbols.find(symbol);
+
+  // If the functions for monitorenter or monitorexit do not exist
+  // (this can only happen if the java-models libary was not downloaded).
+  // We implement them as skips because symex would crash, since it cannot find
```

Comment:
Any such "crash" (whatever that is) would be a bug and that needs to be fixed. I'm confused.

Reply:
This is badly worded. If the methods "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V" and "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V"  are not found (this happens  when the java-core-models-library is  not loaded and the method in question is synchronized).  Symex will rightfully complain as it cannot find the symbols associated with function calls to the aforementioned methods (Note: we prevent JBMC from stubbing these methods). To avoid this and thereby ensuring JBMC works with and without  the models,  we simply such calls with skips when the  java-core -models-library  is not loaded.

Reply:
Will reword comment to make this clear.

Reply:
I'll do a PR that adds these symbols.

Reply:
@kroening In your PR (https://github.com/diffblue/cbmc/pull/2286/) you are unconditionally adding the symbols for monitorenter/exit, while what's needed is adding them *only* when core models cannot be found. Note: that the actual locking mechanism is implemented in these models and as such those models are not simply empty stubs.